Documentation update (sprint 3) (#34)
blahajjessie/kakaw: api-docs/interface/server-socket.md

- Leaderboard
    - `name` (string) : the player's username 
    - `score` (number) : the score of the player

# Types

These are all defined in `respTypes.ts` with 
- a class that contains the name and data
- an interface that contains the data

These messages are the same for players and the host unless specified.

Possible types (all sent from server to client unless otherwise specified):

## `end`

Whoever sends this message is closing the connection for the specified reason. E.g. the server could send {“type”: “end”, “reason”: “That game does not exist”}

Fields: 
- reason (string)

## `beginQuestion`

Sent by server whenever a new question is available

Fields: 
- `questionText`: The text for the question
- `answerTexts`: The answer texts (same as in JSON)
- `time` (number): The number of *miliseconds* that are left in the question (where 0 is the end of the question)
- `score` (number): player’s current score, the host will recieve garbage
- `username`: the name of the player. 
future:

> - `questionIndex` : the index of the question started.

Reviewer: I'd rather have this alongside all the other fields maybe with a note "backend does not send this currently" or something, instead of a "future" section. As frontend is looking for questionIndex right now, not in the future, and isn't going to work without it.

## `endQuestion`

Sent by server when the timer on a question runs out or the host clicks “End now,” and moves clients to the screen to review results
The same message is sent to the host, but the information about the answer choice may be inacurate. 

Fields: 
- `correctAnswers` (numeric array) : 
- `score` (number) = the player’s current score, the host will recieve garbage
- `scoreChange` (number) = how much their score increased due to this question, The host will recieve garbage
- `correct` (boolean) : if the player's answer to the question is correct. The host will recieve garbage
- `leaderboard`: sorted array of `Leaderboard` objects. 
- `time` : number: the ammount of time the player took to answer the question

future:

> - `questionIndex` : the index of the question started.

Reviewer: See comment on lines 45-46

## `playerAction`

future: 

Every time a player completes an action (join, answer) the host is sent this message
It will contain a player id and username that has completed the most recent action. 

Fields: 
- `playerId`: The playerid who completed the action

> - `username`: the username of the player who complteted the action

Reviewer 2: Typo
```suggestion
- `username`: the username of the player who completed the action
```

### Data Structs:

- Every message is a JSON object with type as a string, and potentially other fields depending on the type.

- When a client connects to the server and the game has already started, the server immediately sends a question or endQuestion (or lobby if client is the host) message to catch them up to the current state of the game. Before then, the client just displays “Waiting for the host to start the game” or something. This way the same logic handles both “client joined and is waiting for game to start” and “client joined late and is waiting for the server to respond to them.”

# Types

These messages are the same for players and the host unless specified.

Possible types (all sent from server to client unless otherwise specified):

## `end`

Whoever sends this message is closing the connection for the specified reason. E.g. the server could send {“type”: “end”, “reason”: “That game does not exist”}

Fields: 
- reason (string)

> ## `startQuestion`

Reviewer: Needs questionIndex

Author: that isnt currently sent by the backend.
i can add that as a todo

Sent by server whenever a new question is available

Fields: 
- `questionText`: The text for the question
- `answerTexts`: The answer texts (same as in JSON)
- `time` (number): The number of *miliseconds* that are left in the question (where 0 is the end of the question)
- `score` (number): player’s current score, the host will recieve garbage
- `username`: the name of the player. 

> ## `endQuestion`

Reviewer: Should include username too

Reviewer: And questionText/answerTexts. Also clarify what goes to which clients. Do players get the leaderboard? I don't think so

Reviewer: And questionIndex

Sent by server when the timer on a question runs out or the host clicks “End now,” and moves clients to the screen to review results
The same message is sent to the host too except with only the correctAnswers field (maybe a leaderboard later).

Fields: 
- `correctAnswers` (numeric array) : 
- `score` (number) = the player’s current score, the host will recieve garbage
- `scoreChange` (number) = how much their score increased due to this question, The host will recieve garbage
- `correct` (boolean) : if the player's answer to the question is correct. The host will recieve garbage
- `leaderboard`: sorted array of [{name: string, score: number }, …]

## `playerAction`

Every time a player completes an action (join, answer) the host is sent this message
It will contain a player id and username that has completed the most recent action.

> This may be sent many times rapidly on the case of a host just conneccting. 

Reviewer: We don't need to send it many times, we can send one message with all the actions so far.

Fields: 

> - `player`: {`id`, `username`}

Reviewer: Should that be a colon if this is a map?

Reviewer: Also should we rename this to `players`?

# Types

These messages are the same for players and the host unless specified.

Possible types (all sent from server to client unless otherwise specified):

## `end`

Whoever sends this message is closing the connection for the specified reason. E.g. the server could send {“type”: “end”, “reason”: “That game does not exist”}

Fields: 
- reason (string)

## `startQuestion`

Sent by server whenever a new question is available

Fields: 
- `questionText`: The text for the question
- `answerTexts`: The answer texts (same as in JSON)
- `time` (number): The number of *miliseconds* that are left in the question (where 0 is the end of the question)
- `score` (number): player’s current score
- `username`: the name of the player. 

## `endQuestion`

Sent by server when the timer on a question runs out or the host clicks “End now,” and moves clients to the screen to review results
The same message is sent to the host too except with only the correctAnswers field (maybe a leaderboard later).

Fields: 

> - `correct` (boolean) : if the question is correct

Reviewer: This is presumably sent only to users and it is whether _their answer_ is correct, not the question itself?

- `time` (number): The number of *miliseconds* that are left in the question (where 0 is the end of the question)
- `score` (number): player’s current score, the host will recieve garbage
- `username`: the name of the player. 

> future:
> - `questionIndex` : the index of the question started.

Reviewer: ```suggestion
- `index` : the index of the question started.
```

## `endQuestion`

Sent by server when the timer on a question runs out or the host clicks “End now,” and moves clients to the screen to review results
The same message is sent to the host, but the information about the answer choice may be inacurate. 

Fields: 
- `correctAnswers` (numeric array) : 
- `score` (number) = the player’s current score, the host will recieve garbage
- `scoreChange` (number) = how much their score increased due to this question, The host will recieve garbage
- `correct` (boolean) : if the player's answer to the question is correct. The host will recieve garbage
- `leaderboard`: sorted array of `Leaderboard` objects. 

> - `time` : number: the ammount of time the player took to answer the question

Reviewer: ```suggestion
- `responseTime` : number: the amount of time in milliseconds the player took to answer the question
```

> 
> future:
> - `questionIndex` : the index of the question started.

Reviewer: ```suggestion
- `index` : the index of the question started.
- `username`: the name of the player
```

## `playerAction`

future: 

Every time a player completes an action (join, answer) the host is sent this message

> It will contain a player id and username that has completed the most recent action. 

Reviewer: ```suggestion
It will contain a player id and username that has completed the most recent action. It is also sent to the host right after they join with all the players that have done that action.
```

Fields: 

> - `playerId`: The playerid who completed the action
> - `username`: the username of the player who completed the action

Reviewer: ```suggestion
- `players`: map from ID to username of all the players (possibly just one).
```
